### DISCUSSION TITLE
Is there anyway to get and put several OPCUA nodes in asynchronous way?

### DISCUSSION THREAD

**Discussion by heritopo** [FreeOpcUa/opcua-asyncio#754] [Q&A]
Hello,
I'm a beginner user here.
I have an issue regarding retrieve OPCUA values and return them in json format.
Currently, I have a script like this which only take a value in a time. I also give a small explaination regarding the script with "#".

>import requests
import json
import pandas
from opcua import Client
import time

# 'Get' command to download the desired submodel
>resp_get = requests.get(
    'http://localhost:51310/aas/RobotStation/submodels/EnergyManagement/elements/rActivePowerTotal_Robot')

# 'Get' command response
>response = {200: 'Sucessful',
            204: 'No Content',
            301: 'Moved Permenantly',
            400: 'Bad Request',
            401: 'Unauthorized',
            403: 'Forbidden',
            404: 'Not Found',
            500: 'Internal Server Error'}

>response_code_get = resp_get.status_code
print('The "Get" command response is: {}'.format(response[response_code_get]))

# Saving the submodel as a json format
>submodel_get = resp_get.json()

>with open('AAS.json', 'w') as outfile:
    json.dump(submodel_get, outfile)
>payload=submodel_get["elem"]

#Reading OPCUA data
>while True:
    SERVER_ADDRESS = "opc.tcp://opcua.demo-this.com:51210/UA/SampleServer"
    NODE_ADDRESS = 'ns=2;i=10851'
    client = Client(SERVER_ADDRESS) 
    client.connect()
    node = client.get_node(NODE_ADDRESS) `#this is 'get node' command just for 1 value`
    value= node.get_data_value().Value.Value
    #Adding opcua data to json submode
    payload["value"]= value
    # 'Put' command to upload the edited submodel
    resp_put = requests.put('http://localhost:51310/aas/RobotStation/submodels/EnergyManagement/elements', json=payload) `#this is 'put node' command just for 1 value`
    # 'Put' command response
    response_code_put = resp_put.status_code
    print('The "Put" command response is: {}'.format(response[response_code_put]))
    tick = time.time()
    time.sleep(2)

#Saving the new submodel as a json file
#with open('new.json', 'w') as outfile:
>   json.dump(payload, outfile)`

Do you have any suggestion or idea to get and put several OPCUA nodes in asynchronous way?

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/

i dont think its part of the OPC UA Spec.
tbh it sounds strange what you suggest, the urls are just the namespacenames not really uasable urls for http requests

**Reply by AndreasHeine** [reply]
this is at least what opc ua spec. says about clientside nodemanagement
https://reference.opcfoundation.org/Core/docs/Part1/7.5/
https://reference.opcfoundation.org/Core/docs/Part4/5.7.1/

if you just want to read/write multiple nodes at once the client has methods for that.

**Reply by heritopo** [reply]
@AndreasHeine thank you for your guidance. i forgot to insert the SS of the result when posted it.  Below is the SS of data which i got throught web browser
![image](https://user-images.githubusercontent.com/96148109/146158779-a53a7f60-b769-4399-ae5b-d316eed34789.png)
currently, i can read and upload only a node. I am trying to read multiple node at once

**Reply by AndreasHeine** [reply]
typically for opcua if you want to read multiple there is a list of nodeids or so in the opc ua json request (but thats guessing at this point)!

**Reply by AndreasHeine** [reply]
like in the example @GoetzGoerisch posted: 
if you look at the request
```json
{
    "ServiceId": 629,
    "Body": {
        "RequestHeader": {
            "Timestamp": "2021-12-15T09:54:24.821Z",
            "RequestHandle": 1,
            "TimeoutHint": 30000,
            "ReturnDiagnostics": 2,
            "AuthenticationToken": null
        },
        "MaxAge": 0,
        "TimestampsToReturn": 0,
        "NodesToRead": [
            {
                "NodeId": {
                    "IdType": 1,
                    "Id": "Scalar_Simulation_Double,ns=2",
                    "Namespace": 2
                },
                "AttributeId": 13
            }
        ]
    }
}
```

@heritopo NodesToRead is a list! just expand it ;)

**Reply by heritopo** [reply]
@AndreasHeine thanks a lot. The Example and your explanation make it clear. I will try it

**Comment by swamper123** [comment]
OPC UA itself may can handle HTTPS requests (see [OPC UA HTTPS](https://reference.opcfoundation.org/v104/Core/docs/Part7/6.6.144/), even [HTTPS UA-JSON ](https://reference.opcfoundation.org/v104/Core/docs/Part7/6.6.146/)is in the specs). Haven't seen or dealed with it so far, but in my understanding, it could be possible.

**But** we don't support it at the moment, just TCP binary. What is possible is to read and write data to a node. Even async (via asyncio)! [Take a look in the examples](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-example.py).

**Reply by heritopo** [reply]
@swamper123 thank you for your support. i will  take a look the example

**Comment by GoetzGoerisch** [comment]
The OPC UA REST API Demo can be found [here](https://prototyping.opcfoundation.org/).

**Reply by heritopo** [reply]
@GoetzGoerisch thank you for reference. it will be usefull

**Reply by GoetzGoerisch** [reply]
Sure. Maybe this is also of interest: https://github.com/umati/Dashboard-OPCUA-Client subscribing to OPC UA server and publishing als MQTT to broker. **Note**: Not according UA-JSON encoding.

**Reply by heritopo** [reply]
@GoetzGoerisch i have an issue regarding publishing the data value to server. i am trying  fetch all the node present in the OPCUA server in one time.Currently in simulation of OPC UA server, i have 3 node. 
I am using below code to get(download) data from local host , read the OPC UA data and reupload to server. when acces data through browser, 1 of 3 node is woking. other is failed , the value is not present. 
Do you have any idea about it ? thanks

    import requests
    import json
    import pandas as pd
    from opcua import Client
    import time

    #read excel database submodel & element
    excel= 'D:/aas test/RESTAPI/GET.xlsx'

    Getex= pd.read_excel(excel,sheet_name='GET')  

    #define server address
    url_server='http://localhost:51310/aas/{}/submodels/{}/elements/{}'
    url_server1='http://localhost:51310/aas/{}/submodels/{}/elements/'

    for x in range (len(Getex)):
        aas=Getex.aas[x]
        subm=Getex.submodels[x]
        elem=Getex.elements[x]
        url_serverx=url_server.format(aas,subm,elem)
    # 'Get' command to download the desired submodel
    resp_get = requests.get(url_serverx)
    
    # 'Get' command response
    response = {200: 'Sucessful',
                204: 'No Content',
                301: 'Moved Permenantly',
                400: 'Bad Request',
                401: 'Unauthorized',
                403: 'Forbidden',
                404: 'Not Found',
                500: 'Internal Server Error'}
    
    response_code_get = resp_get.status_code
    print('The "Get" command response is: {}'.format(response[response_code_get]))
    
    # Saving the submodel as a json format
    submodel_get = resp_get.json()
    
    with open('AAS.json', 'w') as outfile:
        json.dump(submodel_get, outfile)
    payload=submodel_get["elem"]
    
    #Reading OPCUA data
    while True:
        sig=Getex.signaladdress[x]  #sig=node id list 
        url_server1x=url_server1.format(aas,subm)
        SERVER_ADDRESS = "opc.tcp://127.0.0.1:4841"
        NODE_ADDRESS = sig
        client = Client(SERVER_ADDRESS)
        client.connect()
        node = client.get_node(NODE_ADDRESS)
        value= node.get_data_value().Value.Value
        #Adding opcua data to json submode
        payload["value"]= value
        # 'Put' command to upload the edited submodel
        resp_put = requests.put(url_server1x, json=payload)
        # 'Put' command response
        response_code_put = resp_put.status_code
        print('The "Put" command response is: {}'.format(response[response_code_put]))
        tick = time.time()
        time.sleep(2)

#Saving the new submodel as a json file
#with open('new.json', 'w') as outfile:
#    json.dump(payload, outfile)
failed ->value is not present
![image](https://user-images.githubusercontent.com/96148109/146460523-c0f6a4dd-6d65-4536-9587-5996fb52136d.png)
succes
![image](https://user-images.githubusercontent.com/96148109/146460589-22cefca6-ecb3-42df-8688-2264a777dea5.png)

**Reply by GoetzGoerisch** [reply]
@heritopo, from the technical side, I hope that @swamper123 and @AndreasHeine  might have a hind?

From a standardization point of view, why do you want to read a AAS submodel from an UA server?
For robots, the OPC 40010-1 UA for Robotics is the information model to use.
For the energy monitoring. the working group is working on have an harmonized approach for all sectors soon.

**Reply by swamper123** [reply]
@GoetzGoerisch Haven't done much in that field, sorry. :confused:

**Reply by heritopo** [reply]
@GoetzGoerisch thank you for your response. The submodel contain information from AASX explore which is as data storage. so we need download AAS in to server. Through OPC-UA,the value will be read and it's put  in to submodel. Then the data's value can be accesed through web browser.

**Reply by heritopo** [reply]
@GoetzGoerisch i don't have any idea about OPC 40010-1 UA. do you have any reference ?

**Reply by GoetzGoerisch** [reply]
> @GoetzGoerisch i don't have any idea about OPC 40010-1 UA. do you have any reference ?

Sure https://reference.opcfoundation.org/Robotics/docs/

@xenonforlife, can answer some questions.
